core/hotspot/cache/wtinylfu/tinylfu.go

```diff
@@ -0,0 +1,223 @@
+//          |                    `---------'                  |
+// .-------------------.              |    .------------------.
+// | Window Cache (1%) |              |    | Main Cache (99%) |
+// |      (LRU)        |              |    |      (SLRU)      |
+// `-------------------'              |    `------------------'
+//          ^                         |               ^
+//          |                         `---------------'
+//       new item                        Winner
+type TinyLfu struct {
+	countMinSketch *countMinSketch
+	doorkeeper     *doorkeeper
+	additions      int
+	samples        int
+	lru            *lru
+	slru           *slru
+	items          map[interface{}]*list.Element
+	stats          *stats.CacheStats
+}
+
+func NewTinyLfu(cap int, isRecordingStats bool) (*TinyLfu, error) {
+	if cap <= 0 {
+		return nil, errors.New("Must provide a positive size")
+	}
+	if cap < 100 {
+		cap = 100
+	}
+	lruCap := int(float64(cap) * lruRatio)
+	slruSize := cap - lruCap
+	var statsCache *stats.CacheStats
+	if isRecordingStats {
+		statsCache = stats.NewCacheStats()
+	}
+	items := make(map[interface{}]*list.Element)
+	return &TinyLfu{
+		countMinSketch: newCountMinSketch(countersFactor * cap),
+		additions:      0,
+		samples:        samplesFactor * cap,
+		doorkeeper:     newDoorkeeper(doorkeeperFactor*cap, falsePositiveProbability),
+		items:          items,
+		lru:            newLRU(lruCap, items),
+		slru:           newSLRU(slruSize, items),
+		stats:          statsCache,
+	}, nil
+}
+
+// Get looks up a key's value from the cache.
+func (t *TinyLfu) Get(key interface{}) (interface{}, bool) {
+	return t.get(key, false)
+}
+
+func (t *TinyLfu) get(key interface{}, isInternal bool) (interface{}, bool) {
+	t.additions++
+	if t.additions == t.samples {
+		t.countMinSketch.reset()
+		t.doorkeeper.reset()
+		t.additions = 0
+	}
+
+	val, ok := t.items[key]
+	if !ok {
+		keyHash := sum(key)
+		if t.doorkeeper.put(keyHash) {
+			t.countMinSketch.add(keyHash)
+		}
+		if !isInternal && t.stats != nil {
+			t.stats.RecordMisses()
+		}
+		return nil, false
+	}
+	item := val.Value.(*slruItem)
+	if t.doorkeeper.put(item.keyHash) {
+		t.countMinSketch.add(item.keyHash)
+	}
+
+	v := item.value
+	if item.listId == admissionWindow {
+		t.lru.get(val)
+	} else {
+		t.slru.get(val)
+	}
+	if !isInternal && t.stats != nil {
+		t.stats.RecordHits()
+	}
+	return v, true
+}
+
+// Contains checks if a key is in the cache without updating
+func (t *TinyLfu) Contains(key interface{}) (ok bool) {
+	_, ok = t.items[key]
+	return ok
+}
+
+func (t *TinyLfu) Add(key interface{}, val interface{}) {
+	t.AddIfAbsent(key, val)
+}
+
+// AddIfAbsent adds item only if key is not existed.
+func (t *TinyLfu) AddIfAbsent(key interface{}, val interface{}) (priorValue interface{}) {
+
+	// Check for existing item
+	if v, ok := t.get(key, true); ok {
+		return v
+	}
+
+	newItem := slruItem{admissionWindow, key, val, sum(key)}
+	candidate, evicted := t.lru.add(newItem)
```

Comment:
The following logic have multi operation of `TinyLfu.data`, Maybe we need optimize logic.

Reply:
The performance of map is much higher than that of linked list operations, and more map operations are used to optimize linked list performance. For performance, map value is an element of the linked list. So every time you manipulate the linked list, you need to re-assign the map value, and the map operation cannot be optimized.